src/types.ts

```diff
@@ -1,6 +1,46 @@
 import * as revalidateEvents from './constants/revalidate-events'
 
-export type Fetcher<Data> = (...args: any) => Data | Promise<Data>
+export type Result<T = unknown> = T | Promise<T>
+
+export type Fetcher<Data = unknown, Args extends Key = Key> =
+  /**
+   * () => [{ foo: string }, { bar: number }] | null
+   *
+   * () => ( [{ foo: string }, { bar: number } ] as const | null )
+   */
+  Args extends (() => readonly [...infer K] | null)
+    ? ((...args: [...K]) => Result<Data>)
+      /**
+       * [{ foo: string }, { bar: number } ] | null
+       *
+       * [{ foo: string }, { bar: number } ] as const | null
+       */
+    : Args extends (readonly [...infer K])
+    ? ((...args: [...K]) => Result<Data>)
+      /**
+       * () => string | null
+       * () => Record<any, any> | null
+       */
+    : Args extends (() => infer T | null)
+    ? (...args: [T]) => Result<Data>
+      /**
+       *  string | null
+       */
+    : Args extends (string | null)
+      /**
+       * when key is Record<any, any> | null
+       * use { foo: string, bar: number } | null as example
+       *
+       * the fetcher would be
+       * (arg0: string) => any | (arg0: { foo: string, bar: number }) => any
+       * so we add this condition to make (arg0: string) => any to be never
+       */
+    ? Args extends (Record<any, any> | null)
+      ? never
+      : (...args: [string]) => Result<Data>
+    : Args extends (infer T)
+    ? (...args: [T]) => Result<Data>
```

Comment:
Can those 2 cases (string, record) be merged as one?

```ts
Args extends (infer T)
  ? (...args: [T]) => Result<Data>
  : never
```

Reply:
The simple answer is no.

But i just realize it could have a batter implementation. 
## TLDR

```typescript
// When conditional types act on a generic type, they become distributive when given a union type
// So if we have type like.
type Fn<Args extends (string | null | Record<any,any>) = string | null | Record<any,any>> = 
  Args extends (infer T)
    ? (...args: [T]) => any
    : never

// the result here is actully a union type 
// (string extends (infer T) ? (...args: [T] =>  any) : never) | (null extend (infer T) ? (...args: [T] =>  any) : never)  
// so the result is (args: string) => any | (args: null) => any
type Test = Fn<string | null>


// we have to give it more specific condition
type Fn2<Args extends (string | null | Record<any,any>) = string | null | Record<any,any>> = 
  Args extends (string | null)
    ? (...args: [string]) => any
    : Args extends (infer T)
    ? (...args: [T]) => any
    : never

// Now These would work
type Test2 = Fn2<string>
type Test3 = Fn2<string | null>
type Test4 = Fn<Record<string, number>>

// (args: null) => any | (args: Record<string, number>) => any
type Test5 = Fn<Record<string, number> | null>


// add condition to remove (args: null) => any
type Fn3<Args extends (string | null | Record<any,any>) = string | null | Record<any,any>> = Args extends null
      ? never
      : Args extends (infer T)
    ? (...args: [T]) => any
    : never

type Test6 = Fn3<string>
type Test7 = Fn3<string | null>
type Test8 = Fn3<Record<string, number>>
type Test9 = Fn3<Record<string, number> | null>
```

[Playground](https://www.typescriptlang.org/play?#code/PTAEHUAsFMDtQMYHtYBMCWAXdKCGAbUTATwAdoBnUXBTUFa0AczmgCd0EizoAaImMVAAjaMgC20UBgqYOwgK7YAblIDuMeE3Sr4uUAtg54JcgCgQoAMpJQ6AGag1UyLlXdyofOgDW0AHRmplIAYrAAPACCbExU0AAemHCoVAAUshywTKAAPqCwCviEeQBKYkhsqOG4sMS8NcQAfACUoAC8oBnoWbn5hcWgZciV1bX1tY2N7aBmoKDRsaAJSWhp3fbsoAAqzbNzoAD8oKn+p7gxFABcoADaWwC6rW1TDXtz17DQqmxmFmCYMFAbEohToMGBdioNEw-SE+kMxg8Uj+xy6PWWyTWsA2bG2rSOJzOF2ud3u7Sm1FqrQ+X3YrTyqQKRSWiWSx3Wmx2h2Op3851iJIe5LmDWp+VpbFaM0sFFsAKkwIooMhx35V06cm6TCeL1qvVSao+-R1lOIQR420odA6YXCaOyeSZ+EavxRzlArncmFs2ncWFA4gqUgo5AQDk4iBQGGwKHNnjCACYohcWSsUqjNT1Hf1ekMKlUGuMmk8NZkHX1maVyiNCw1JtM9gs4qzVhmy70nbt9tzCXzibd7Y9ha9u9cm6nMezsZyu-sCbzDbcHiaR-sad9fpYAHJINTbGAUdRIQqoJwVHxxqRbK0J6aJu2Zpgu4KW2QAZjvsCT9o7-WfFuvWQABZP3CPMRntfgCnEUQ2EmTcwANftO2HPUGUXcCqkgvoYPYFpULNF9AMwABWUDMIfMsoIUXC4N-IoXQQ6hUFPZA0CwRFvSBaBA3cJCBQrfAV1qS9QDCN9k0WDFW3SR96IGCjawmEsf2zStBmrAsxjrKYOnHaT0ydN453FDdu3eeYUwMrEcTxYz5yJATSWEs1R1M9hfiIq0ADZPwk+1-08YiAHY-MorV5MCq8rQADjCijsOg2D4K82QAE54s08Ksmo2ipjU50gA)
[Distributive Conditional Types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types)

Reply:
Great idea! 👍